### PR TITLE
Undo poc

### PR DESCRIPTION
**Pull request by emord**
blah

@dimagi/product going to put this on staging. Expecting this needs a trip through UAT?

cc: @snopoke

### CODE REVIEW

**Comment by amsagoff** [comment]
Indeed - lets uat!

File: src/mugs.js [hunk 1]
```diff
@@ -475,6 +475,17 @@ define([
...
     });
 
+    Object.defineProperty(Mug.prototype, "previousSibling", {
+        get: function () {
+            var tree = this.form.tree,
+                node = this.form.tree.getNodeFromMug(this),
+                parentMug = this.parentMug,
+                parentNode = parentMug ? tree.getNodeFromMug(parentMug) : tree.getRootNode(),
+                mugPosition = parentNode.children.indexOf(node);
+            return (mugPosition === 0) ? parentMug : parentNode.children[mugPosition - 1].getValue();
```
Comment (millerdev) [inline src/mugs.js]: What do you think of delegating to the tree? There's a lot of tree implementation details in here. This getter would then `return this.form.tree.getPreviousSibling(this)`

File: src/undomanager.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    $,
+    _,
+    UNDO_ALERT
+) {
+    function alertShown() {
+        return $('.fd-undo-delete').length;
+    }
+
+    function createAlert() {
+        $('.fd-scrollable-tree').prepend(UNDO_ALERT);
```
Comment (millerdev) [inline src/undomanager.js]: I think the location of this alert might be problematic in large forms where the user has scrolled the question tree.

**Comment by millerdev** [comment]
excellent!

**Comment by czue** [comment]
super cool!

File: src/mugs.js (hunk 1 above)
Comment (millerdev) [inline src/mugs.js]: Also, consider adding a docstring to explain what "previous sibling" means in the context of

- a question whose previous sibling is it's parent
- the first question in the tree
- any other non-obvious details

File: src/undomanager.js (hunk 2 above)
Comment (emord) [inline src/undomanager.js]: https://github.com/dimagi/Vellum/commit/908e94aeaceb9ddeb22f291e2915b0b6277aa69f

**Comment by millerdev** [comment]
Looking good. Should we add *awaiting QA* to this until UAT is finished? Or were you planning to deploy it behind a flag?

**Comment by emord** [comment]
Passed QA http://manage.dimagi.com/default.asp?217694

**Comment by emord** [comment]
:tada:

**Comment by czue** [comment]
:recycle:

**Pull request merged**